rclcpp/test/test_wait_set.cpp

```diff
@@ -0,0 +1,280 @@
+        std::vector<rclcpp::GuardCondition::SharedPtr>{},
+        std::vector<rclcpp::TimerBase::SharedPtr>{},
+        std::vector<rclcpp::ClientBase::SharedPtr>{},
+        std::vector<rclcpp::ServiceBase::SharedPtr>{},
+        std::vector<rclcpp::WaitSet::WaitableEntry>{},
+        context);
+      (void)wait_set;
+    }, rclcpp::exceptions::RCLInvalidArgument);
+  }
+}
+
+/*
+ * Testing rcl wait set accessor.
+ */
+TEST_F(TestWaitSet, get_rcl_wait_set) {
+  {
+    rclcpp::WaitSet wait_set;
+    wait_set.get_rcl_wait_set();
+  }
+}
+
+/*
+ * Testing add/remove for guard condition methods.
+ */
+TEST_F(TestWaitSet, add_remove_guard_condition) {
+  // normal, mixed initialization
+  {
+    auto gc = std::make_shared<rclcpp::GuardCondition>();
+    auto gc2 = std::make_shared<rclcpp::GuardCondition>();
+    rclcpp::WaitSet wait_set({}, {gc});
+    wait_set.add_guard_condition(gc2);
+    wait_set.remove_guard_condition(gc2);
+    wait_set.remove_guard_condition(gc);
+  }
+
+  // out of order removal
+  {
+    auto gc = std::make_shared<rclcpp::GuardCondition>();
+    auto gc2 = std::make_shared<rclcpp::GuardCondition>();
+    rclcpp::WaitSet wait_set({}, {gc});
+    wait_set.add_guard_condition(gc2);
+    wait_set.remove_guard_condition(gc);
+    wait_set.remove_guard_condition(gc2);
+  }
+
+  // start empty, normal
+  {
+    auto gc = std::make_shared<rclcpp::GuardCondition>();
+    rclcpp::WaitSet wait_set;
+    wait_set.add_guard_condition(gc);
+    wait_set.remove_guard_condition(gc);
+  }
+
+  // add invalid (nullptr)
+  {
+    rclcpp::WaitSet wait_set;
+    ASSERT_THROW(
+    {
+      wait_set.add_guard_condition(nullptr);
+    }, std::invalid_argument);
+  }
+
+  // double add
+  {
+    auto gc = std::make_shared<rclcpp::GuardCondition>();
+    rclcpp::WaitSet wait_set;
+    wait_set.add_guard_condition(gc);
+    ASSERT_THROW(
+    {
+      wait_set.add_guard_condition(gc);
+    }, std::runtime_error);
+  }
+
+  // remove invalid (nullptr)
+  {
+    rclcpp::WaitSet wait_set;
+    ASSERT_THROW(
+    {
+      wait_set.remove_guard_condition(nullptr);
+    }, std::invalid_argument);
+  }
+
+  // remove unrelated
+  {
+    auto gc = std::make_shared<rclcpp::GuardCondition>();
+    auto gc2 = std::make_shared<rclcpp::GuardCondition>();
+    rclcpp::WaitSet wait_set({}, {gc});
+    ASSERT_THROW(
+    {
+      wait_set.remove_guard_condition(gc2);
+    }, std::runtime_error);
+  }
+
+  // double remove
+  {
+    auto gc = std::make_shared<rclcpp::GuardCondition>();
+    rclcpp::WaitSet wait_set({}, {gc});
+    wait_set.remove_guard_condition(gc);
+    ASSERT_THROW(
+    {
+      wait_set.remove_guard_condition(gc);
+    }, std::runtime_error);
+  }
+
+  // remove from empty
+  {
+    auto gc = std::make_shared<rclcpp::GuardCondition>();
+    rclcpp::WaitSet wait_set;
+    ASSERT_THROW(
+    {
+      wait_set.remove_guard_condition(gc);
+    }, std::runtime_error);
+  }
+
+  // remove after delete, checking weak ownership behavior
```

Comment:
I think you should uncomment this test case, or remove it if it's not something we can really test.

Reply:
I'll delete it then I guess, it is a case I thought would be a corner case but for an non-obvious reason wasn't actually a problem. I guess I was leaving it there so in the future if someone also thought that should be a test they might see this and know we've thought about it already, but I guess that's not likely.

Reply:
f0dcf15